File: python/Maze.py
########################################################
# You have got a maze, which is a n*n Grid. Every cell
# of the maze contains these numbers 1, 2 or 3. If it
# contains 1: means we can go Right from that cell only.
# If it contains 2: means we can go Down from that cell
# only. If it contains 3: means we can go Right and Down
# to both paths from that cell. We cant go out of the maze
# at any time. Initially, You are on the Top Left Corner
# of The maze(Entry). And, You need to go to the Bottom
# Right Corner of the Maze(Exit). You need to find the
# total number of paths from Entry to Exit Point. There
# may be many paths but you need to select that path which
# contains the maximum number of Adventure.
# The Adventure on a path is calculated by taking the sum
# of all the cell values thatlies in the path.
#
# Input: matrix = { {1,1,3,2,1},
#                   {3,2,2,1,2},
#                   {1,3,3,1,3},
#                   {1,2,3,1,2},
#                   {1,1,1,3,1}}
# Output: {4,18}
# Explanation: There are total 4 Paths Available 
# out of which The Max Adventure is 18. Total 
# possible Adventures are 18,17,17,16. Of these 
# 18 is the maximum.
#
# Expected Time Complexity: O(n^2)
# Expected Space Complexity: O(n^2)
#  
########################################################
def wtf( p): return( p%1000000007)
RYTE, DOWN = 1, 2
class Solution:
    def FindWays( self, maze): #########################
        n = len( maze)
        def cons(): return [[0]* n for _ in range( n)]
        # setup adventures and paths fst row and col
        adv = cons(); adv[0][0] = maze[0][0]
        pts = cons(); pts[0][0] = 1
        for j in range( 1, n):
            if maze[0][j-1] == DOWN: break
            adv[0][j] = adv[0][j-1] + maze[0][j]
            pts[0][j] = 1
        for i in range( 1, n):
            if maze[i-1][0] == RYTE: break
            adv[i][0] = adv[i-1][0] + maze[i][0]
            pts[i][0] = 1
########################################################
#         +-------+
#         | i-1,j | adv[i, j] = max(adv[up], adv[left])
# +-------+-------+ pts[i, j] = pts[up] + pts[left], if
# | i,j-1 |  i,j  | allowed
# +-------+-------+
########################################################
        for i in range( 1, n):
            for j in range( 1, n):
                if maze[i-1][j] != RYTE and pts[i-1][j]:
                    adv[i][j] = adv[i-1][j] + maze[i][j]
                    pts[i][j] = pts[i-1][j]
                if maze[i][j-1] != DOWN and pts[i][j-1]:
                    adv[i][j] = max( adv[i][j], adv[i][j-1] + maze[i][j])
                    pts[i][j] += pts[i][j-1]
        if not pts[-1][-1]: return( 0, 0)
        return( wtf( pts[-1][-1]), adv[-1][-1])
########################################################
if __name__ == '__main__':
    T = int( input())
    for i in range( T):
        n = int( input())
        matrix = []
        for _ in range( n):
            cur = list( map( int, input().split()))
            matrix.append( cur)
        obj = Solution()
        ans = obj.FindWays( matrix)
        for _ in ans:
            print( _, end = " ")
        print()
########################################################
# log: Test Cases Passed: 230 / 230
#      Total Time Taken: 0.21 / 1.24

File: cpp/maze.cpp
////////////////////////////////////////////////////////////////
// Rat in a Maze Problem - I
////////////////////////////////////////////////////////////////
// Consider a rat placed at (0,0) in a square matrix of order 
// N*N. It has to reach the destination at (N-1,N-1). Find all 
// possible paths that the rat can take to reach from source to 
// destination. The directions in which the rat can move are 
// 'U' (up), 'D' (down), 'L' (left), 'R' (right). Value 0 at a 
// cell in the matrix represents that it is blocked and rat 
// cannot move to it while value 1 at a cell in the matrix 
// represents that rat can be travel through it.
////////////////////////////////////////////////////////////////
// Note: In a path, no cell can be visited more than one time.
// If the source cell is 0, the rat cannot move to any other 
// cell.
////////////////////////////////////////////////////////////////
// Example: 1
////////////////////////////////////////////////////////////////
// Input: N = 4
//        m[][] = {{1, 0, 0, 0},
//                 {1, 1, 0, 1},
//                 {1, 1, 0, 0},
//                 {0, 1, 1, 1}}
// Output: DDRDRR DRDDRR
// Explanation: nop
////////////////////////////////////////////////////////////////
// Expected Time Complexity: O((3N^2)).
// Expected Auxiliary Space: O(L*X), L = length of the path, 
// X = number of paths.
////////////////////////////////////////////////////////////////
# include <iostream>
# include <vector>
# include <string>
# include <algorithm>
////////////////////////////////////////////////////////////////
using namespace std;
////////////////////////////////////////////////////////////////
class Solution {
    static constexpr int NFDR{ 4 }; // number of directions
    static constexpr int GRD{ 2 }; // guard
    vector<string> pathList; // output
    string path{ "T" }; // dropped from the top initially
    vector<int> maze; // 32-bit flat model( with protection )
    int n{}; // maze's width, height
    vector<int> dr{ NFDR }; // URDL directions displacement
    int goal{}; // target offset
    void Dfs( int src );
    void Debug() const;
    void Flatten( const vector<vector<int>>& m );
public:
    vector<string> findPath( vector<vector<int>>& m, int n );
};
////////////////////////////////////////////////////////////////
void Solution::Flatten( const vector<vector<int>>& m ){
    const vector<int> grd( n, GRD );
    maze.insert( maze.end(), grd.begin(), grd.end());
    for( const auto& row: m ){
        maze.push_back( GRD );
        maze.insert( maze.end(), row.begin(), row.end());
        maze.push_back( GRD );
    }
    maze.insert( maze.end(), grd.begin(), grd.end());
}
vector<string> Solution::findPath
( vector<vector<int>>& m, int n ){
    if( m[ 0 ][ 0 ] == 0 ){
        return {};
    }
//  @Override
    m[ n - 1 ][ n - 1 ] = 1;
    this->n = n + 2; // protection enabled
    // One advantage of using flat array than 2D is that moving
    // up down, etc. requires only one instead of two operations.
    dr = { -this->n, 1, this->n, -1 };
    maze.reserve( this->n * this->n );
    Flatten( m );
    goal = ( int )maze.size() - this->n - 2;
    Debug();
    Dfs( this->n + 1 );
    return pathList;
}
void Solution::Dfs( int src ){
    if( src == goal ){
        pathList.push_back
        ( string( path.begin() + 1, path.end()));
    } else {
        maze[ src ] = GRD;
        for( int j{}; j < NFDR; ++j ){
            int dst{ src + dr[ j ]};
            if( maze[ dst ] == 1 ){
                path.push_back( "URDL"[ j ]);
                Dfs( dst );
            }
        }
        maze[ src ] = 1;
    }
    path.pop_back();
}
void Solution::Debug() const {
    cout << goal << endl;
    for( int j{}; j < maze.size(); ++j ){
        cout << maze[ j ] << " ";
        if(( j + 1 ) % n == 0 ){
            cout << endl;
        }
    }
}
////////////////////////////////////////////////////////////////
// Test Cases Passed: 1          6     2/ 1         6    2
// Total Points Scored:                          4/          4
// Your Total Score:       1              0  9              8
// Total Time Taken:               0 . 0               1
//////////////////////////////|///////////////|///////////////|/
// T           h i          s |    i        s |       t   h e |
///////////////|///////////////////|/////////////|//////////////
// n        th |   t  i         me |    I  '   m |  so l  v in g
////////////////////////|//////////////////////////////|////////
// s  i m     i l    ar |  p            r o b   le   m |
////////////////////////////////////////////////////////////////
int main(){
    int n;
    cin >> n;
    vector<vector<int>> m( n, vector<int>( n, 0 ));
    for( int i = 0; i < n; i++ ){
        for( int j = 0; j < n; j++ ){
            cin >> m[ i ][ j ];
        }
    }
    Solution obj;
    vector<string> result = obj.findPath( m, n );
    sort( result.begin(), result.end());
    if( result.size() == 0 ){
        cout << -1;
    } else {
        for( int i = 0; i < result.size(); i++ ){ 
            cout << result[i] << " ";
        }
    }
}
////////////////////////////////////////////////////////////////